hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NSSummaryEndpoint.java

```diff
@@ -0,0 +1,640 @@
+      totalCnt += getTotalDirCount(subdir);
+    }
+    return totalCnt;
+  }
+
+  /**
+   * Given an object ID, return total data size (no replication)
+   * under this object.
+   * @param objectId the object's ID
+   * @return total used data size in bytes
+   * @throws IOException ioEx
+   */
+  private long getTotalSize(long objectId) throws IOException {
+    NSSummary nsSummary = reconNamespaceSummaryManager.getNSSummary(objectId);
+    if (nsSummary == null) {
+      return 0L;
+    }
+    long totalSize = nsSummary.getSizeOfFiles();
+    for (long childId: nsSummary.getChildDir()) {
+      totalSize += getTotalSize(childId);
+    }
+    return totalSize;
+  }
+
+  /**
+   * Given an object ID, return the file size distribution.
+   * @param objectId the object's ID
+   * @return int array indicating file size distribution
+   * @throws IOException ioEx
+   */
+  private int[] getTotalFileSizeDist(long objectId) throws IOException {
+    NSSummary nsSummary = reconNamespaceSummaryManager.getNSSummary(objectId);
+    if (nsSummary == null) {
+      return new int[ReconConstants.NUM_OF_BINS];
+    }
+    int[] res = nsSummary.getFileSizeBucket();
+    for (long childId: nsSummary.getChildDir()) {
+      int[] subDirFileSizeDist = getTotalFileSizeDist(childId);
+      for (int i = 0; i < ReconConstants.NUM_OF_BINS; ++i) {
+        res[i] += subDirFileSizeDist[i];
+      }
+    }
+    return res;
+  }
+
+  private List<OmBucketInfo> listBucketsUnderVolume(final String volumeName)
+          throws IOException {
+    List<OmBucketInfo> result = new ArrayList<>();
+    if (Strings.isNullOrEmpty(volumeName)
+            || !checkVolumeExistence(volumeName)) {
+      return null;
+    }
+    Table bucketTable = omMetadataManager.getBucketTable();
+    String seekPrefix =
+            omMetadataManager.getVolumeKey(volumeName + OM_KEY_PREFIX);
+
+    TableIterator<String, ? extends Table.KeyValue<String, OmBucketInfo>>
+            iterator = bucketTable.iterator();
+
+
+    while (iterator.hasNext()) {
+      Table.KeyValue<String, OmBucketInfo> kv = iterator.next();
+
+      String key = kv.getKey();
+      OmBucketInfo omBucketInfo = kv.getValue();
+
+      if (omBucketInfo != null) {
+        // We should return only the keys, whose keys match with the seek prefix
+        if (key.startsWith(seekPrefix)) {
+          result.add(omBucketInfo);
+        }
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Helper function to check if a path is a directory, key, or invalid.
+   * @param volName volume name
+   * @param bucketName bucket name
+   * @param keyName key name
+   * @return DIRECTORY, KEY, or UNKNOWN
+   * @throws IOException
+   */
+  private EntityType determineKeyPath(String volName, String bucketName,
+                                      String keyName, long bucketObjectId)
+          throws IOException {
+    omMetadataManager.getLock().acquireReadLock(BUCKET_LOCK, volName,
+            bucketName);
```

Comment:
I would argue the bucket lock here is unnecessary. As we don't ever write to the OM DB snapshot on the Recon side.